pylate/losses/cached_contrastive.py

```diff
@@ -0,0 +1,348 @@
+from __future__ import annotations
+
+from collections.abc import Iterator
+from contextlib import nullcontext
+from functools import partial
+from typing import Iterable, Optional
+
+import torch
+import torch.nn.functional as F
+import tqdm
+from torch import Tensor, nn
+from torch.utils.checkpoint import get_device_states, set_device_states
+
+from ..models import ColBERT
+from ..scores import colbert_scores
+from .contrastive import extract_skiplist_mask
+
+
+class RandContext:
+    """
+    Random-state context manager class. Reference: https://github.com/luyug/GradCache.
+
+    This class will back up the pytorch's random state during initialization. Then when the context is activated,
+    the class will set up the random state with the backed-up one.
+    """
+
+    def __init__(self, *tensors) -> None:
+        self.fwd_cpu_state = torch.get_rng_state()
+        self.fwd_gpu_devices, self.fwd_gpu_states = get_device_states(*tensors)
+
+    def __enter__(self) -> None:
+        self._fork = torch.random.fork_rng(devices=self.fwd_gpu_devices, enabled=True)
+        self._fork.__enter__()
+        torch.set_rng_state(self.fwd_cpu_state)
+        set_device_states(self.fwd_gpu_devices, self.fwd_gpu_states)
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self._fork.__exit__(exc_type, exc_val, exc_tb)
+        self._fork = None
+
+
+def _backward_hook(
+    grad_output: Tensor,
+    sentence_features: Iterable[dict[str, Tensor]],
+    loss_obj,
+) -> None:
+    """
+    A backward hook that re-runs the forward for each mini-batch with gradients enabled
+    and uses the cached partial derivatives w.r.t. the embeddings to backprop.
+    """
+    assert loss_obj.cache is not None
+    assert loss_obj.random_states is not None
+    with torch.enable_grad():
+        for sentence_feature, grad, random_states in zip(
+            sentence_features, loss_obj.cache, loss_obj.random_states
+        ):
+            for (reps_mb, _), grad_mb in zip(
+                loss_obj.embed_minibatch_iter(
+                    sentence_feature=sentence_feature,
+                    with_grad=True,
+                    copy_random_state=False,
+                    random_states=random_states,
+                ),
+                grad,
+            ):
+                # Dot the embedding chunk with the cached gradient chunk,
+                # scaled by grad_output from the top-level backward pass
+                surrogate = (
+                    torch.dot(reps_mb.flatten(), grad_mb.flatten()) * grad_output
+                )
+                surrogate.backward()
+
+
+class CachedContrastive(nn.Module):
+    """
```

Comment:
Description should start on the same line as the quote """ for documentation parsing